[SPARK-29213][SQL] Generate extra IsNotNull predicate in FilterExec

wangshuo128 opened the pull request:
### What changes were proposed in this pull request?
Currently the behavior of getting output and generating null checks in `FilterExec` is different. Thus some nullable attribute could be treated as not nullable by mistake.

In `FilterExec.ouput`, an attribute is marked as nullable or not by finding its `exprId` in notNullAttributes:
```
a.nullable && notNullAttributes.contains(a.exprId)
```
But in `FilterExec.doConsume`,  a `nullCheck` is generated or not for a predicate is decided by whether there is semantic equal not null predicate:
```
      val nullChecks = c.references.map { r =>
        val idx = notNullPreds.indexWhere { n => n.asInstanceOf[IsNotNull].child.semanticEquals(r)}
        if (idx != -1 && !generatedIsNotNullChecks(idx)) {
          generatedIsNotNullChecks(idx) = true
          // Use the child's output. The nullability is what the child produced.
          genPredicate(notNullPreds(idx), input, child.output)
        } else {
          ""
        }
      }.mkString("\n").trim
```
NPE will happen when run the SQL below:
```
sql("create table table1(x string)")
sql("create table table2(x bigint)")
sql("create table table3(x string)")
sql("insert into table2 select null as x")
sql(
  """
    |select t1.x
    |from (
    |    select x from table1) t1
    |left join (
    |    select x from (
    |        select x from table2
    |        union all
    |        select substr(x,5) x from table3
    |    ) a
    |    where length(x)>0
    |) t3
    |on t1.x=t3.x
  """.stripMargin).collect()
```
NPE Exception:
```
java.lang.NullPointerException
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage2.processNext(generated.java:40)
    at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
    at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:726)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
    at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:135)
    at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
    at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:94)
    at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
    at org.apache.spark.scheduler.Task.run(Task.scala:127)
    at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:449)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:452)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
```
the generated code:
```
== Subtree 4 / 5 ==
*(2) Project [cast(x#7L as string) AS x#9]
+- *(2) Filter ((length(cast(x#7L as string)) > 0) AND isnotnull(cast(x#7L as string)))
   +- Scan hive default.table2 [x#7L], HiveTableRelation `default`.`table2`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [x#7L]

Generated code:
/* 001 */ public Object generate(Object[] references) {
/* 002 */   return new GeneratedIteratorForCodegenStage2(references);
/* 003 */ }
/* 004 */
/* 005 */ // codegenStageId=2
/* 006 */ final class GeneratedIteratorForCodegenStage2 extends org.apache.spark.sql.execution.BufferedRowIterator {
/* 007 */   private Object[] references;
/* 008 */   private scala.collection.Iterator[] inputs;
/* 009 */   private scala.collection.Iterator inputadapter_input_0;
/* 010 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[] filter_mutableStateArray_0 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[2];
/* 011 */
/* 012 */   public GeneratedIteratorForCodegenStage2(Object[] references) {
/* 013 */     this.references = references;
/* 014 */   }
/* 015 */
/* 016 */   public void init(int index, scala.collection.Iterator[] inputs) {
/* 017 */     partitionIndex = index;
/* 018 */     this.inputs = inputs;
/* 019 */     inputadapter_input_0 = inputs[0];
/* 020 */     filter_mutableStateArray_0[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(1, 0);
/* 021 */     filter_mutableStateArray_0[1] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(1, 32);
/* 022 */
/* 023 */   }
/* 024 */
/* 025 */   protected void processNext() throws java.io.IOException {
/* 026 */     while ( inputadapter_input_0.hasNext()) {
/* 027 */       InternalRow inputadapter_row_0 = (InternalRow) inputadapter_input_0.next();
/* 028 */
/* 029 */       do {
/* 030 */         boolean inputadapter_isNull_0 = inputadapter_row_0.isNullAt(0);
/* 031 */         long inputadapter_value_0 = inputadapter_isNull_0 ?
/* 032 */         -1L : (inputadapter_row_0.getLong(0));
/* 033 */
/* 034 */         boolean filter_isNull_2 = inputadapter_isNull_0;
/* 035 */         UTF8String filter_value_2 = null;
/* 036 */         if (!inputadapter_isNull_0) {
/* 037 */           filter_value_2 = UTF8String.fromString(String.valueOf(inputadapter_value_0));
/* 038 */         }
/* 039 */         int filter_value_1 = -1;
/* 040 */         filter_value_1 = (filter_value_2).numChars();
/* 041 */
/* 042 */         boolean filter_value_0 = false;
/* 043 */         filter_value_0 = filter_value_1 > 0;
/* 044 */         if (!filter_value_0) continue;
/* 045 */
/* 046 */         boolean filter_isNull_6 = inputadapter_isNull_0;
/* 047 */         UTF8String filter_value_6 = null;
/* 048 */         if (!inputadapter_isNull_0) {
/* 049 */           filter_value_6 = UTF8String.fromString(String.valueOf(inputadapter_value_0));
/* 050 */         }
/* 051 */         if (!(!filter_isNull_6)) continue;
/* 052 */
/* 053 */         ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
/* 054 */
/* 055 */         boolean project_isNull_0 = false;
/* 056 */         UTF8String project_value_0 = null;
/* 057 */         if (!false) {
/* 058 */           project_value_0 = UTF8String.fromString(String.valueOf(inputadapter_value_0));
/* 059 */         }
/* 060 */         filter_mutableStateArray_0[1].reset();
/* 061 */
/* 062 */         filter_mutableStateArray_0[1].zeroOutNullBytes();
/* 063 */
/* 064 */         if (project_isNull_0) {
/* 065 */           filter_mutableStateArray_0[1].setNullAt(0);
/* 066 */         } else {
/* 067 */           filter_mutableStateArray_0[1].write(0, project_value_0);
/* 068 */         }
/* 069 */         append((filter_mutableStateArray_0[1].getRow()));
/* 070 */
/* 071 */       } while(false);
/* 072 */       if (shouldStop()) return;
/* 073 */     }
/* 074 */   }
/* 075 */
/* 076 */ }

```

This PR proposes to use semantic comparison both in `FilterExec.output` and `FilterExec.doConsume` for nullable attribute.

With this PR, the generated code snippet is below:
```
== Subtree 2 / 5 ==
*(3) Project [substring(x#8, 5, 2147483647) AS x#5]
+- *(3) Filter ((length(substring(x#8, 5, 2147483647)) > 0) AND isnotnull(substring(x#8, 5, 2147483647)))
   +- Scan hive default.table3 [x#8], HiveTableRelation `default`.`table3`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [x#8]

Generated code:
/* 001 */ public Object generate(Object[] references) {
/* 002 */   return new GeneratedIteratorForCodegenStage3(references);
/* 003 */ }
/* 004 */
/* 005 */ // codegenStageId=3
/* 006 */ final class GeneratedIteratorForCodegenStage3 extends org.apache.spark.sql.execution.BufferedRowIterator {
/* 007 */   private Object[] references;
/* 008 */   private scala.collection.Iterator[] inputs;
/* 009 */   private scala.collection.Iterator inputadapter_input_0;
/* 010 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[] filter_mutableStateArray_0 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[2];
/* 011 */
/* 012 */   public GeneratedIteratorForCodegenStage3(Object[] references) {
/* 013 */     this.references = references;
/* 014 */   }
/* 015 */
/* 016 */   public void init(int index, scala.collection.Iterator[] inputs) {
/* 017 */     partitionIndex = index;
/* 018 */     this.inputs = inputs;
/* 019 */     inputadapter_input_0 = inputs[0];
/* 020 */     filter_mutableStateArray_0[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(1, 32);
/* 021 */     filter_mutableStateArray_0[1] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(1, 32);
/* 022 */
/* 023 */   }
/* 024 */
/* 025 */   protected void processNext() throws java.io.IOException {
/* 026 */     while ( inputadapter_input_0.hasNext()) {
/* 027 */       InternalRow inputadapter_row_0 = (InternalRow) inputadapter_input_0.next();
/* 028 */
/* 029 */       do {
/* 030 */         boolean inputadapter_isNull_0 = inputadapter_row_0.isNullAt(0);
/* 031 */         UTF8String inputadapter_value_0 = inputadapter_isNull_0 ?
/* 032 */         null : (inputadapter_row_0.getUTF8String(0));
/* 033 */
/* 034 */         boolean filter_isNull_0 = true;
/* 035 */         boolean filter_value_0 = false;
/* 036 */         boolean filter_isNull_2 = true;
/* 037 */         UTF8String filter_value_2 = null;
/* 038 */
/* 039 */         if (!inputadapter_isNull_0) {
/* 040 */           filter_isNull_2 = false; // resultCode could change nullability.
/* 041 */           filter_value_2 = inputadapter_value_0.substringSQL(5, 2147483647);
/* 042 */
/* 043 */         }
/* 044 */         boolean filter_isNull_1 = filter_isNull_2;
/* 045 */         int filter_value_1 = -1;
/* 046 */
/* 047 */         if (!filter_isNull_2) {
/* 048 */           filter_value_1 = (filter_value_2).numChars();
/* 049 */         }
/* 050 */         if (!filter_isNull_1) {
/* 051 */           filter_isNull_0 = false; // resultCode could change nullability.
/* 052 */           filter_value_0 = filter_value_1 > 0;
/* 053 */
/* 054 */         }
/* 055 */         if (filter_isNull_0 || !filter_value_0) continue;
/* 056 */         boolean filter_isNull_8 = true;
/* 057 */         UTF8String filter_value_8 = null;
/* 058 */
/* 059 */         if (!inputadapter_isNull_0) {
/* 060 */           filter_isNull_8 = false; // resultCode could change nullability.
/* 061 */           filter_value_8 = inputadapter_value_0.substringSQL(5, 2147483647);
/* 062 */
/* 063 */         }
/* 064 */         if (!(!filter_isNull_8)) continue;
/* 065 */
/* 066 */         ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
/* 067 */
/* 068 */         boolean project_isNull_0 = true;
/* 069 */         UTF8String project_value_0 = null;
/* 070 */
/* 071 */         if (!inputadapter_isNull_0) {
/* 072 */           project_isNull_0 = false; // resultCode could change nullability.
/* 073 */           project_value_0 = inputadapter_value_0.substringSQL(5, 2147483647);
/* 074 */
/* 075 */         }
/* 076 */         filter_mutableStateArray_0[1].reset();
/* 077 */
/* 078 */         filter_mutableStateArray_0[1].zeroOutNullBytes();
/* 079 */
/* 080 */         if (project_isNull_0) {
/* 081 */           filter_mutableStateArray_0[1].setNullAt(0);
/* 082 */         } else {
/* 083 */           filter_mutableStateArray_0[1].write(0, project_value_0);
/* 084 */         }
/* 085 */         append((filter_mutableStateArray_0[1].getRow()));
/* 086 */
/* 087 */       } while(false);
/* 088 */       if (shouldStop()) return;
/* 089 */     }
/* 090 */   }
/* 091 */
/* 092 */ }
```
### Why are the changes needed?
Fix NPE bug in FilterExec.

### Does this PR introduce any user-facing change?
no

### How was this patch tested?
new UT

wangshuo128 committed:
[SPARK-29213] Make it consistent when get notnull output and generate null checks in FilteExec

AmplabJenkins commented:
Can one of the admins verify this patch?

wangshuo128 commented:
@viirya @cloud-fan  gentle ping, would you like to review this, thank you. :)

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala:
```diff
@@ -117,7 +117,7 @@ case class FilterExec(condition: Expression, child: SparkPlan)
   override def output: Seq[Attribute] = {
     child.output.map { a =>
-      if (a.nullable && notNullAttributes.contains(a.exprId)) {
+      if (a.nullable && notNullPreds.exists(_.semanticEquals(a))) {
```
Can you explain what the real difference? AFAIK `Attribute#sementicEquals` just compare the expr id.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16306/
Test PASSed.

SparkQA commented:
**[Test build #111232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111232/testReport)** for PR 25902 at commit [`7e54ccb`](https://github.com/apache/spark/commit/7e54ccb54d61f3fce1202714ad5abd366d4ab9cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111232/
Test FAILed.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala:
```diff
@@ -117,7 +117,7 @@ case class FilterExec(condition: Expression, child: SparkPlan)
   override def output: Seq[Attribute] = {
     child.output.map { a =>
-      if (a.nullable && notNullAttributes.contains(a.exprId)) {
+      if (a.nullable && notNullPreds.exists(_.semanticEquals(a))) {
```
notNullPreds is a list of IsNotNull expressions. I think this semanticEquals comparison will fail at all?

wangshuo128 replied:
yeah, this will fail all. This change is not good enough. I'll try another way.

wangshuo128 committed:
remove exprId both in notNullPreds and otherPreds when get notNullAttributes

SparkQA commented:
**[Test build #111255 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111255/testReport)** for PR 25902 at commit [`af0fb3f`](https://github.com/apache/spark/commit/af0fb3f00bba6a445690fe8a7f1a1779b4a5a43b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16324/
Test PASSed.

wangshuo128 commented:
@cloud-fan @viirya 
Sorry, my earlier description and  fix is not good enough .

The real problem is here:
`nullcheck` will not be generated for `(length(cast(x#7L as string)) > 0)` because 
`(length(cast(x#7L as string)) > 0) ` and `isnotnull(cast(x#7L as string))` are not semantic equal when  generating code for `Filter ((length(cast(x#7L as string)) > 0) AND isnotnull(cast(x#7L as string)))`. 
At the same time, filter output  attribute `x#7L` is marked as not nullable. Nullability will not be checked when generating code for `x#7L`.
So NPE is thrown  when reading null data in generated code of  `(length(cast(x#7L as string)) > 0) `.

To fix this, I think we can filter attributes both in `notNullPreds` and  `otherPreds` when get `notNullAttributes`:
```
  private val notNullAttributes = notNullPreds.flatMap(_.references).distinct.map(_.exprId)
    .diff(otherPreds.flatMap(_.references).distinct.map(_.exprId))
```

wangshuo128 committed:
modify test table names

SparkQA commented:
**[Test build #111256 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111256/testReport)** for PR 25902 at commit [`bc0a688`](https://github.com/apache/spark/commit/bc0a68811bdec4990b2fb11e2d953a138d13b61e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16325/
Test PASSed.

SparkQA commented:
**[Test build #111255 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111255/testReport)** for PR 25902 at commit [`af0fb3f`](https://github.com/apache/spark/commit/af0fb3f00bba6a445690fe8a7f1a1779b4a5a43b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111255/
Test FAILed.

SparkQA commented:
**[Test build #111256 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111256/testReport)** for PR 25902 at commit [`bc0a688`](https://github.com/apache/spark/commit/bc0a68811bdec4990b2fb11e2d953a138d13b61e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111256/
Test FAILed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala:
```diff
@@ -110,6 +110,7 @@ case class FilterExec(condition: Expression, child: SparkPlan)
 
   // The columns that will filtered out by `IsNotNull` could be considered as not nullable.
   private val notNullAttributes = notNullPreds.flatMap(_.references).distinct.map(_.exprId)
+    .diff(otherPreds.flatMap(_.references).distinct.map(_.exprId))
```
This fix seems sub-optimal to me. If we have a `IsNotNull(a)`, then `a` should be a `notNullAttribute` even if `a` appears in `otherPreds` like `EqualTo(a, b)`.

cloud-fan replied:
I think the bug is about how we codegen null checks, the current logic is:
1. split the predicates into `notNullPreds` (IsNotNull expressions) and `otherPreds`
2. try codegen `otherPreds` first, then `notNullPreds`.
3. if an other-predicate can leverage a specific not-null-predicate, then codegen that not-null-predicate first, so that we can codegen the other-predicate with non-nullable attributes.

There is a problem in step 3. Image that we have a `IsNotNull(SubString(a))` and a `SomeFunc(a)`. Then `a` is a not-null attribute for sure, even if there is no `IsNullNull(a)` predicate.

When we codegen `SomeFunc(a)`, we can't find a `IsNotNull(a)` expression, so we skip the null check. However, we assume `a` is not nullable when codegen `SomeFunc(a)`, which is wrong as `IsNotNull(SubString(a))` has not been codegened yet.

viirya commented:
I think the problem is because the child of IsNotNull is not a reference. So when we look at the IsNotNull that can filter null values for a non-IsNotNull predicate, we fail to look at it up:

```scala
val idx = notNullPreds.indexWhere { n => n.asInstanceOf[IsNotNull].child.semanticEquals(r)}
```

Doesn't the easiest fix look:
```scala
val idx = notNullPreds.indexWhere { n =>
  n.asInstanceOf[IsNotNull].child.references.contains(r)
}
```

cloud-fan commented:
@viirya your proposal works, but is a little against the original design goal. Think about `IsNotNull(a + b + c + d ..)` and `Substring(a)`.

When codegen `Substring(a)`, with your proposal, we will codegen `IsNotNull(a + b + c + d ..)` first, which loads more columns and may cause perf regression.

I have a slightly different proposal: when codegen `Substring(a)`, check if there is a `IsNotNull(a)` predicate. If there is, codegen it first. If not, and `a` is a not-nullable attribute, then generate a `IsNotNull(a)` and codegen it first.

viirya commented:
Ah, I see. You are correct. Your proposal sounds good.

wangshuo128 commented:
@cloud-fan 
> I have a slightly different proposal: when codegen Substring(a), check if there is a IsNotNull(a) predicate. If there is, codegen it first. If not, and a is a not-nullable attribute, then generate a IsNotNull(a) and codegen it first.

`Substring(a)` a could be an expression. If we have predicates like `Substring(Substring(a))` and `IsNotNulll(Substring(a))`,  we could codegen `IsNotNulll(Substring(a))` for null check first?

cloud-fan commented:
@wangshuo128 you are right that we can improve the algorithm further, and we can do it once the bug is fixed. For now let's fix the bug without introducing a regression.

That said, for any predicate (not `IsNotNull`), find all its references (e.g. `a` and `b`), and check if there are corresponding `IsNotNull` predicates (e.g. `IsNotNull(a)` and `IsNotNull(b)`). If a reference has a corresponding `IsNotNull` predicate, codegen it first. Otherwise, generate a new `IsNotNull` predicate for it and codegen it, if the reference is a not-nullable attribute.

wangshuo128 commented:
> For now let's fix the bug without introducing a regression.

Ok, I'll try to fix this bug first.

wangshuo128 committed:
Some slight modifications so we no longer depend on Hive.
This is from https://github.com/JoshRosen/spark/commit/d1658bbffb07d2b0c5f3aabc9362b397a2c0aeb8

wangshuo128 committed:
Generate extra IsNotNull predicate for null check

SparkQA commented:
**[Test build #111341 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111341/testReport)** for PR 25902 at commit [`5555731`](https://github.com/apache/spark/commit/5555731b3f20cb802995de424e4de0dab90407f8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16388/
Test PASSed.

SparkQA commented:
**[Test build #111341 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111341/testReport)** for PR 25902 at commit [`5555731`](https://github.com/apache/spark/commit/5555731b3f20cb802995de424e4de0dab90407f8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111341/
Test FAILed.

wangshuo128 commented:
Sorry, @JoshRosen , do you mind I use [your code](https://github.com/JoshRosen/spark/commit/d1658bbffb07d2b0c5f3aabc9362b397a2c0aeb8) to move test from sql/hive to sql/core ?

JoshRosen commented:
> Sorry, @JoshRosen , do you mind I use [your code](https://github.com/JoshRosen/spark/commit/d1658bbffb07d2b0c5f3aabc9362b397a2c0aeb8) to move test from sql/hive to sql/core ?

Yes, please feel free to use that! This bug is somewhat hard to reproduce in unit tests because filter pushdown to data source scans will prevent the `FilterExec` code from processing the `null`-containing input rows. In d1658bbffb07d2b0c5f3aabc9362b397a2c0aeb8 I worked round that by using typed Dataset operations to create black-box operators through which the optimizer cannot push filters.

(I'm following this PR because I'm interested in other aspects of `IsNotNull` optimizations; I have some notes on this particular code in #24765, something of a backburner project as I learn more about this code)

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala:
```diff
@@ -171,13 +172,23 @@ case class FilterExec(condition: Expression, child: SparkPlan)
     // TODO: revisit this. We can consider reordering predicates as well.
     val generatedIsNotNullChecks = new Array[Boolean](notNullPreds.length)
+
+    val extraIsNotNullReferences = mutable.Set[Attribute]()
```
nit: `extraIsNotNullAttrs`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala:
```diff
@@ -171,13 +172,23 @@ case class FilterExec(condition: Expression, child: SparkPlan)
+
+    val extraIsNotNullReferences = mutable.Set[Attribute]()
+
+    def outputContainsNotNull(ref: Attribute): Boolean = {
```
This doesn't mean the attribute is not nullable. We should check `notNullAttributes.contains(ref.exprId)`

wangshuo128 replied:
Ah, my mistake.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -3192,6 +3192,33 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
     }
   }
+
+  test("SPARK-29213 Make it consistent when get notnull output and generate null " +
```
We can simply describe the bug in the test name: `SPARK-29213: FilterExec should not throw NPE`

wangshuo128 committed:
Address some comments

SparkQA commented:
**[Test build #111377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111377/testReport)** for PR 25902 at commit [`5272097`](https://github.com/apache/spark/commit/52720971986285a40e83ecdd50c16528db38df65).

wangshuo128 commented:
> Yes, please feel free to use that! 

@JoshRosen Thanks a lot!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16420/
Test PASSed.

wangshuo128 committed:
Code sytle modification

SparkQA commented:
**[Test build #111378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111378/testReport)** for PR 25902 at commit [`fcada1c`](https://github.com/apache/spark/commit/fcada1cd65260601d6b3ef951359dd3ed6b0003d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16421/
Test PASSed.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -3192,6 +3192,32 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
   }
+
+  test("SPARK-29213: FilterExec should not throw NPE") {
+    withView("t1", "t2", "t3") {
```
nit: `withView` -> `withTempView`

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -3192,6 +3192,32 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
+      sql("select ''").as[String].map(identity).toDF("x").createOrReplaceTempView("t3")
+      sql(
+        """
+          |select t1.x
```
nit: No strict rule though, I like capitalized words for SQL keywords: e.g., select -> SELECT, ....

SparkQA commented:
**[Test build #111377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111377/testReport)** for PR 25902 at commit [`5272097`](https://github.com/apache/spark/commit/52720971986285a40e83ecdd50c16528db38df65).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111377/
Test PASSed.

SparkQA commented:
**[Test build #111378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111378/testReport)** for PR 25902 at commit [`fcada1c`](https://github.com/apache/spark/commit/fcada1cd65260601d6b3ef951359dd3ed6b0003d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111378/
Test PASSed.

wangshuo128 committed:
Refine test

SparkQA commented:
**[Test build #111393 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111393/testReport)** for PR 25902 at commit [`f33d34c`](https://github.com/apache/spark/commit/f33d34ced0239ed0158b7d732d62d16e0be79cdb).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16434/
Test PASSed.

SparkQA commented:
**[Test build #111393 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111393/testReport)** for PR 25902 at commit [`f33d34c`](https://github.com/apache/spark/commit/f33d34ced0239ed0158b7d732d62d16e0be79cdb).
 * This patch **fails due to an unknown error code, -9**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111393/
Test FAILed.

maropu commented:
@wangshuo128 Can you resolve the conflict first?

wangshuo128 commented:
@cloud-fan 
Would you like to take another look if you have time? :)
BTW, need I cherry-pick #25925 to resolve the conflicts?

wangshuo128 commented:
> @wangshuo128 Can you resolve the conflict first?

Ok

SparkQA commented:
**[Test build #111460 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111460/testReport)** for PR 25902 at commit [`124ad87`](https://github.com/apache/spark/commit/124ad87721c3e5f1e08f5d2f26cc940d281d79d9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16494/
Test PASSed.

SparkQA commented:
**[Test build #111460 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111460/testReport)** for PR 25902 at commit [`124ad87`](https://github.com/apache/spark/commit/124ad87721c3e5f1e08f5d2f26cc940d281d79d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111460/
Test PASSed.

cloud-fan commented:
thanks, merging to master/2.4!

wangshuo128 commented:
Thanks!
